### ISSUE TITLE
Setup input binding when debugging in vs19

### ISSUE THREAD

**Issue by peirens-bart** [Azure/iotedgehubdev#267]
I'm new to IoT development, and I'm creating a generic module so I just need to add 1 dll to the container image for changing the behaviour for the whole module.
in vs19 it's working except for one part, debugging a single module message input using curl or postman with an input different then "input1"

My code is mapping the method names to a generic message handler, this way I can use method.inputname for invoke the correct method.
`foreach (var method in _moduleManger.MethodNames)
{
                _logger.Info($"Register message callback for {method.ToLowerInvariant()}");
                await _ioTHubModuleClient.SetInputMessageHandlerAsync(method.ToLowerInvariant(), PipeMessageToModule, _ioTHubModuleClient);
}`

When I new press F5 the iotedgehubdev is started in single mode
`...Documents\Visual Studio 2019\iotedgehubdev\v0.13.0\iotedgehubdev start -i "input1"`

the problem is now when i send a message with curl
`$ curl --header "Content-Type: application/json" --request POST --data '{"inputName": "startsendingevent","data":"{\"type\":\"HertzEvent\",\"id\":0,\"Key\":\"StartSendingEvent\",\"sendername\":\"test\",\"eventname\":\"test\",\"hertzevetarg\": []}"}' http://localhost:53000/api/v1/messages
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   202  100    22  100   180     93    765 --:--:-- --:--:-- --e:-h-:--   859{"message":"accepted"}
`
I receive a accpted, but the breakpoint in t handler is never hit, because must start with
`Documents\Visual Studio 2019\iotedgehubdev\v0.13.0\iotedgehubdev start -i "StartSendingEvent"`

Is there a config file that I can add tho my project so the launch settings can change.
I can't see the options for -i in the Tool->Azure Iot edge Tools->IoT edge tools Settings

Or is there an options to use a .iotedgeproj with e deployment where i can do the same thing.
I don't realy want the test the whole routing, just a single module.

**Comment by SLdragon** [comment]
Hi,  @peirens-bart , for current version of Visual Studio extension, it will parse the input name according to your routes settings, for exameple, if your routes settings is as below:

```
        "routes": {
          "IotEdgeModule1ToIoTHub": "FROM /messages/modules/IotEdgeModule1/outputs/* INTO $upstream",
          "sensorToIotEdgeModule1": "FROM /messages/modules/SimulatedTemperatureSensor/outputs/temperatureOutput INTO BrokeredEndpoint(\"/modules/IotEdgeModule1/inputs/input3\")"
        },
```

Then Visual Studio would call iotedgehubdev start -i "input3"

And for your request, I will disccuss with team members and maybe we would add an additonal settings for input name in future release of our extension.

**Comment by peirens-bart** [comment]
@SLdragon thank jou for the answer.

Do you known if there is a solution if I don't have a module yet for the input?
Can  I  bind the same handlerlogic to a methodhandler and use CTD methods for testing?

**Comment by SLdragon** [comment]
Currently, only VSCode edge extension can assign input name manually, if you want to use custom input name, you can try our VSCode extension, and it also very convenient to use.

Take a look and have a try:
https://marketplace.visualstudio.com/items?itemName=vsciot-vscode.azure-iot-edge

**Comment by peirens-bart** [comment]
@SLdragon Ok, I'll wait for now, VsCode is not an option for now.

I've created a workaround with some extra code
in de launchSettings.json I added commandLineArgs --addinput1=true --mapInput1to=startsendingevent

And now like aps.net I created a ModuleHostbuilder
`public static ModuleStarter Build(string[] args)
        {
            var configuration = new ConfigurationBuilder()
                .AddCommandLine(args)
                .Build();
            return new ModuleStarter(configuration);
        }`

I also changed the code of the modulestarter a litle bit
` 
if(_configuartion.GetValue("addinput1", false))
                {
                    await _ioTHubModuleClient.SetInputMessageHandlerAsync("input1", PipeInput1MessageToModule, _ioTHubModuleClient);
                }
            }
        }

        public Task<MessageResponse> PipeInput1MessageToModule(Message message, object userContext)
        {
            var messageName = _configuartion.GetValue("mapInput1to", "input1");
            var hertzEvent = HertzEventSerializer.FromBinary(message.GetBytes());
            hertzEvent.EventName = messageName;
            if (hertzEvent != null)
            {
                _moduleManger.SendHertzEventToModule(hertzEvent);
            }
            return Task.FromResult(MessageResponse.None);
        }`

this is working for now.
I noticed that you added the label feauture request, can this be somthing like this? Adding the value of -i in the launcsettings.json?

**Comment by SLdragon** [comment]
Thank you @peirens-bart for your information, to add config in the launchsettings.json file is a solution, however, it would introduce some extra logic in the code to handle this.

I think the root cause of this problem is that edge module do not have a property to describe the usage of  input name, so we can only infer it from routes settings or ask user to provide it.

For VSCode it is flexible and can use command palette to ask user to provide it before starting debugger, however, this is not a best practice to ask user to provide something before starting debugger in Visual Studio world, so our team need further discussion about this.

Thank you so much for your solution, and this may help other users who also suffering this issue.

**Comment by SLdragon** [comment]
Hi, @peirens-bart, by the way, which circumstance as you said " there is a solution if I don't have a module yet for the input", if there is no module yet, I think visual studio cannot start single module debug, it only works when set a module project as an startup project

**Comment by peirens-bart** [comment]
@SLdragon, It is working. I just need to press F5 when the module is the start up project.

the thing is, input will come from a module that connects to the hardware, this will send the raw input to a processing modules, what can be of any kind. the implementation is behind an interface. so my docker build will copy the correct dll and I have another module.
the module wrapper is the one I want to debug/ test.

see https://docs.microsoft.com/en-us/azure/iot-edge/how-to-visual-studio-develop-module#build-and-debug-single-module

**Comment by SLdragon** [comment]
Thank you @peirens-bart , I know your scenario

**Comment by SLdragon** [comment]
Hi, @peirens-bart, could you also provide some samples about your project, so that we can take a look  to verify if we can support it, thanks!

**Comment by peirens-bart** [comment]
I created a light version
`curl --header "Content-Type: application/json" --request POST --data '{"inputName": "input1","data":"{\"prop1\":1}"}' http://localhost:53000/api/v1/messages`
to invoke the input 1.
it is not possible to use InputEvent1 or InputEvent2 as inputname.
for this ot work the call to 
`iotedgehubdev\v0.13.0\iotedgehubdev start -i "input1"`
need to change
`iotedgehubdev\v0.13.0\iotedgehubdev start -i "input1;InputEvent1;InputEvent2"`

[ConsoleApp.zip](https://github.com/Azure/iotedgehubdev/files/4349329/ConsoleApp.zip)


change the launchsettings.json for testing InputEvent2
```
{
  "profiles": {
    "ConsoleApp": {
      "commandName": "Project",
      "commandLineArgs": "--addinput1=true --mapInput1to=InputEvent1"
    }
  }
}
```

**Comment by SLdragon** [comment]
Thank you peirens-bart, I just tried your sample project and it works fine, I put a full project sample here and maybe others meet this issue can refer:

[SampleProject.zip](https://github.com/Azure/iotedgehubdev/files/4385900/SampleProject.zip)

**Comment by peirens-bart** [comment]
mayby usefull to add envirinment variables in the launchsettings, since the moduleid is not filled in in the device twin properties. even the live version the moduleid is not filled in but you need to use the same veriable.
on azure
![image](https://user-images.githubusercontent.com/18715894/77634837-ebf5a580-6f51-11ea-8aed-050facba41b4.png)

add to launchsettings.json

```
      "environmentVariables": {
        "IOTEDGE_MODULEID": "moduleid"
      }
```

in code do this

`_moduleName = Environment.GetEnvironmentVariable("IOTEDGE_MODULEID")
`
and not
```
Twin iotModuleTwin = await _ioTHubModuleClient.GetTwinAsync();
 _moduleName = iotModuleTwin.ModuleId; // this is always null
```

**Comment by konichi3** [comment]
Hi 

Thank you for filing the issue.

We are looking to clean up the reported issues that are obsolete such that our team can focus on the latest issues reported.
As such, we are going to close the issue for now.

If the issue is still blocking or critical to you, please file a new issue and we will do our best to respond as soon as possible.
When you file a new issue, please ensure following additional information are provided.

o	If you believe it’s a bug, provide detail repro steps. 
o	If it’s a feature request, please note as such
o	Expected outcome
o	Is this a blocking issue?

Thank you,
Azure IoT Developer and Device team
